content/docs/userspace/gall-guide/11-fail.md

```diff
@@ -0,0 +1,141 @@
++++
+title = "11. Failure"
+weight = 55
+template = "doc.html"
++++
+
+In this lesson we'll cover the last agent arm we haven't touched on yet:
+`on-fail`. We'll also touch on one last concept, which is the _helper core_.
+
+# Failures
+
+When crashes or errors occur in certain cases, Gall passes them to an agent's
+`on-fail` arm for handling. This arm is very seldom used, almost all agents
+leave it for `default-agent` to handle, which just prints the error message to
+the terminal. While you're unlikely to use this arm, we'll briefly go over its
+behavior for completeness.
+
+`on-fail` takes a `term` error message and a `tang`, typically containing a
+stack trace, and often with additional messages about the error. If it weren't
+delegated to `on-fail:def`, it would begin with:
+
+```hoon
+++  on-fail
+  |=  [=term =tang]
+  ^-  (quip card _this)
+  ....
+```
+
+Gall calls `on-fail` in four cases:
+
+- When there's a crash in the `on-arvo` arm.
+- When there's a crash in the `on-agent` arm.
+- When there's a crash in the `on-leave` arm.
+- When an agent produces a `%watch` card but the `wire`, ship, agent and `path`
+  specified are the same as an existing subscription.
+
+For an `on-arvo` failure, the `term` will always be `%arvo-response`, and the
+`tang` will contain a stack trace.
+
+For `on-agent`, the `term` will be the head of the `sign` (`%poke-ack`, `%fact`,
+etc). The `tang` will contain a stack trace and a message of "closing
+subscription".
+
+For an `on-leave` failure, the `term` will always be `%leave`, and the `tang`
+will contain a stack trace.
+
+For a `%watch` failure, the `term` will be `%watch-not-unique`. The `tang` will
+include a message of "subscribe wire not unique", as well as the agent name, the
+`wire`, the target ship and the target agent.
+
+How you might handle these cases (if you wanted to manually handle them) depends
+on the purpose of your particular agent.
+
+## Helper core
```

Comment:
Feels a little bit weird to have this as part of the page on `+on-fail`. Any particular reason this doesn't get its own page? ie, a "best practices for larger applications" might be nice.

Reply:
the on-fail bit is quite short, and it was information overload to go into helper core stuff when I introduced other stuff like that in lesson 7. It's basically that these two things alone would be a little light on information to be a lesson